### PR TITLE
PWN-9633 - Various optimizations related to user tokens, metadata and extensions

### PR DESCRIPTION
**Pull request by 1sakov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-9633

## Description of Work

1. We got rid of early token metadata requests: 
```
No user token found for mint So11111111111111111111111111111111111111112
No user token found for mint 8HGyAAB1yoM1ttS7pXjHMa3dukTFGQggnFFH3hJZgzQh
No user token found for mint ATLASXmbPQxBUYbxPsV97usA3fPQYEqzQBUHgiFCUsXx
No user token found for mint 8upjSpvjcdpuzhfR1zriwg5NXkwDruejqNE9WNbPRtyA
No user token found for mint SCYfrGCw8aDiqdgcpdGjV6jp4UVVQLuphxTDLNWu36f
No user token found for mint Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB
No user token found for mint MERt85fc5boKw3BW1eYdxonEuJNvXbiMbs6hvheau5K
No user token found for mint MAPS41MDahZ9QdKXhVa4dWB9RuyfV4XqhyAZ8XcYepb
No user token found for mint F6v4wfAdJB8D8p77bMXZgYt8TDKsYxLYxH5AFhUkYx9W
```

2. Removed double check for token `visibility` in repository and dao. Now tokens are being update much faster inside Room
3. Optimised extensions. Now we don't override the token for its extensions
4. Now extensions from metadata are being applied correctly
5. Removed additional token comparison when reading from DB. Now tokens are sorted BEFORE being inserted in DB
6. Cleaned up [UserTokensDatabaseRepository]

Notes: 
Extensions are not working for `SOL`. Need to workaround

## Screenshots (Optional)

| Screenshot #1 | Screenshot #2 | Screenshot #3 |
|:-------------:|:-------------:|:-------------:|
| [Paste Pic 1] | [Paste Pic 2] | [Paste Pic 3] |

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/updates/handler/SolanaAccountUpdateHandler.kt [hunk 1]
```diff
@@ -30,12 +26,9 @@ class SolanaAccountUpdateHandler(
             .result
             .value
 
-        tokensRepository.updateUserToken(
-            newBalanceLamports = response.lamports,
-            mintAddress = Constants.WRAPPED_SOL_MINT.toBase58Instance(),
-            publicKey = tokenKeyProvider.publicKey.toBase58Instance()
-        )
-        Timber.tag(TAG).d("New SOL balance received, amountInLamports=${response.lamports}, data = $data")
+        userTokensInteractor.updateSolanaBalance(newBalanceInLamports = response.lamports)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/updates/handler/SolanaAccountUpdateHandler.kt:29]: updateSolBalance / updateSolTokenBalance

File: app/src/main/java/org/p2p/wallet/user/interactor/UserTokensInteractor.kt [hunk 2]
```diff
@@ -35,13 +44,58 @@ class UserTokensInteractor(
...
-            oldTokens?.visibility?.let { newToken.copy(visibility = it) }
-            newToken
-        }
+        tokens
+            .map { newToken ->
+                // saving visibility state which user could change by this moment
+                val oldToken = cachedTokens.find { oldTokens -> oldTokens.publicKey == newToken.publicKey }
+                oldToken?.visibility?.let { newToken.copy(visibility = it) }
+                newToken
+            }
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/user/interactor/UserTokensInteractor.kt:98]:
```suggestion
            .updateVisibilityState()
```

File: app/src/main/java/org/p2p/wallet/user/interactor/UserTokensInteractor.kt [hunk 3]
```diff
@@ -72,14 +126,28 @@ class UserTokensInteractor(
...
     suspend fun setTokenHidden(mintAddress: String, visibility: String) =
         userTokensLocalRepository.setTokenHidden(mintAddress, visibility)
+
+    private suspend fun createNewToken(
+        mintAddress: Base58String,
+        newBalanceLamports: BigInteger,
+        accountPublicKey: Base58String
+    ): Token.Active? {
+
+        val tokenMetadataData = userLocalRepository.findTokenData(mintAddress.base58Value) ?: return null
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/user/interactor/UserTokensInteractor.kt:138]: can we have a task to rename UserLocalRepository?

File: app/src/main/java/org/p2p/wallet/user/repository/UserTokensDatabaseRepository.kt [hunk 4]
```diff
@@ -1,51 +1,32 @@
...
-            ?.let { createUpdatedToken(it, newBalanceLamports) }
-            ?: createNewToken(mintAddress, newBalanceLamports, publicKey)
-
-        if (tokenToUpdate != null) {
-            tokensDao.insertOrUpdate(
-                entities = listOf(tokenConverter.toDatabase(tokenToUpdate))
-            )
-        }
+    override suspend fun updateOrCreateUserToken(tokenToUpdate: Token.Active) {
+//        val tokenToUpdate = findTokenByMintAddress(mintAddress)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/user/repository/UserTokensDatabaseRepository.kt]: ?

File: core/src/main/java/org/p2p/core/token/TokenExtensions.kt [hunk 5]
```diff
@@ -19,6 +19,8 @@ data class TokenExtensions(
     val canTokenBeHidden: Boolean? = null
 ) : Parcelable {
 
+    var isRateExceedsTheDifference: Boolean = true
```
Comment (gslevinkov) [inline core/src/main/java/org/p2p/core/token/TokenExtensions.kt:22]: ?

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/home/analytics/BrowseAnalytics.kt [hunk 6]
```diff
@@ -1,5 +1,6 @@
 package org.p2p.wallet.home.analytics
 
+import kotlinx.coroutines.launch
```
[the remark this answers is not in the document]
Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/home/analytics/BrowseAnalytics.kt:3]: why? 
It's by alphabet, should be ok

File: app/src/main/java/org/p2p/wallet/updates/handler/SolanaAccountUpdateHandler.kt (hunk 1 above)
Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/updates/handler/SolanaAccountUpdateHandler.kt:29]: Why you don't like Solana but Sol?

File: app/src/main/java/org/p2p/wallet/user/interactor/TokenMetadataInteractor.kt [hunk 7]
```diff
@@ -20,7 +20,8 @@ class TokenMetadataInteractor(
 
     suspend fun loadAllTokensMetadata() {
         val metadataFromFile: TokensMetadataInfo? = readTokensMetadataFromFile()
-        val modifiedSince = metadataFromFile?.timestamp
+//        val modifiedSince = metadataFromFile?.timestamp
+        val modifiedSince = "Fri, 4 Aug 2023 04:13:27 +0000"
```
[the remark this answers is not in the document]
Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/user/interactor/TokenMetadataInteractor.kt]: already

File: app/src/main/java/org/p2p/wallet/home/analytics/BrowseAnalytics.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/home/analytics/BrowseAnalytics.kt:3]: Screenshot 2023-08-10 at 18 06 35
we do have import layout rules

File: app/src/main/java/org/p2p/wallet/updates/handler/SolanaAccountUpdateHandler.kt (hunk 1 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/updates/handler/SolanaAccountUpdateHandler.kt:29]: we have findSolOrThrow and some other places i can't remember. and also seems to me Solana is more like a network, and Sol is more like a token. 
and Sol also is shorter, and i know you love shorter names :)

File: app/src/main/java/org/p2p/wallet/user/interactor/UserTokensInteractor.kt (hunk 3 above)
Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/user/interactor/UserTokensInteractor.kt:133]: https://p2pvalidator.atlassian.net/browse/PWN-9649
Added todo inside the repo

File: core/src/main/java/org/p2p/core/token/TokenExtensions.kt (hunk 5 above)
Comment (kabramovp2p) [reply core/src/main/java/org/p2p/core/token/TokenExtensions.kt:25]: =[

**Review by eduardmaximovich — APPROVED** [review]

**Pull request merged**